design/authenticators/authn_k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,255 @@
+
+### Cert injection implementation
+
+Now that we decided that we will not wait for the injection, and clarified the response code, we can explore the different options that we have to inject the cert. 
+
+**Note**: The SPDY protocol [is being deprecated](https://github.com/kubernetes/enhancements/issues/384)  so it's not an option.
+
+#### Option 1 - Run a parallel request to verify cert injection
+In this option, we send the tar request as we do today, but also create a parallel WebSocket in which we will send another request to verify that the cert was successfully injected. We have several options for the verification command:
+1. Run an "ls" command on the cert path. This way we get an immediate feedback on the  cert presence. However, this may return a positive answer even if the cert is not fully injected (this assumption will be investigated as part of this effort).
+1. Calculate the SHA-1 hash of the cert content before sending it and then sending  a request to get the calculated SHA-1 hash of the cert in the container. Once the two hashes are equal then we know that the cert was fully installed.
+
+Note: we use `SHA-1` and not `MD5` as the latter is not FIPS compliant. 
+
+#### Option 2 - Send the cert content as part of the URL
+In this option, we do not use STDIN to send the cert but send it in the URL. This eliminates the bug in KubeCtl Exec as without STDIN we will get the response and would be able to know whether the injection succeeded or not, and log the outcome.
+
+However, we may have a limit on the size of the URL and we will have an issue  if the certificate will grow in the future. To overcome this, we can split the  request. The first one will create the file. The following requests will populate it with the cert content. Then the last request will verify that the whole content has passed (with one of the methods described above). We can decide on the number  of requests dynamically according to the size of the cert.
+
+We have a security concern with this option. While the traffic is fully encrypted when using https, including the URL, [it's still possible that the URL will be visible in certain locations](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), while the request travels to its destination. It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.
+
+Thus, we will not use this option due to security concerns.
+
+#### Option 3 - Client sets its State once cert is injected
+This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code).
```

Comment:
```suggestion
This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code)).
```

Reply:
also, has this been updated since the earlier section was updated? I felt when reading to this point that this was already something you were setting out to do

Reply:
this wasn't updated but i left this as an option so people know it was considered. I'll add a tldr in the beginning of the section that option 4 was chosen in case someone wants to skip.